shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/ExaminationAttributes.java

```diff
@@ -0,0 +1,129 @@
+                }
+            }
+        }
+        return res;
+	}
+
+	public void addDatasetAttributes(T acquisitionId, T datasetId, Attributes attributes) {
+		if (!acquisitionMap.containsKey(acquisitionId)) {
+            acquisitionMap.put(acquisitionId, Optional.of(new AcquisitionAttributes<T>()));
+        }
+        acquisitionMap.get(acquisitionId).get().addDatasetAttributes(datasetId, attributes);
+	}
+
+    public static void addDatasetAttributes(ExaminationAttributes<Long> examinationAttributes, Examination examination, Attributes singleImageAttributes) {
+        //String serieUID = singleImageAttributes.getString(Tag.SeriesInstanceUID);
+        String sopUID = singleImageAttributes.getString(Tag.SOPInstanceUID);
+        if (sopUID != null && examination != null && examination.getDatasetAcquisitions() != null && examinationAttributes != null) {
+            for (DatasetAcquisition acquisition : examination.getDatasetAcquisitions()) {
+                if (acquisition.getDatasets() != null) {
+                    for (Dataset dataset : acquisition.getDatasets()) {
+                        if (dataset.getDatasetExpressions() != null) {
+                            for (DatasetExpression expression : dataset.getDatasetExpressions()) {
+                                if (expression.getDatasetFiles() != null) {
+                                    for (DatasetFile file : expression.getDatasetFiles()) {
+                                        if (file.getPath() != null) {
+                                            String datasetSopUID =  WADODownloaderService.extractInstanceUID(file.getPath());
+                                            if (sopUID.equals(datasetSopUID)) {
+                                                examinationAttributes.addDatasetAttributes(acquisition.getId(), dataset.getId(), singleImageAttributes);
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        for (T acqId : acquisitionMap.keySet()) {
+            sb.append("acquisition ").append(acqId).append("\n");
+            for(String line : acquisitionMap.get(acqId).toString().split("\n")) {
+                sb.append("\t").append(line).append("\n");
+            }
+        }
+        return sb.toString();
+    }
+
+    public Set<T> getAcquisitionIds() {
+        return acquisitionMap.keySet();
+    }
+
+    public void addAcquisitionAttributes(T acquisitionId, AcquisitionAttributes<T> dicomAcquisitionAttributes) {
+        if (acquisitionMap.containsKey(acquisitionId) && acquisitionMap.get(acquisitionId).isPresent()) {
+            acquisitionMap.get(acquisitionId).get().merge(dicomAcquisitionAttributes);
```

Comment:
Hi @julien-louis, this code looks somehow strange, why do you merge acquisitionAttributes?

Reply:
Here I want to add acquisition dicom attributes to the index acquisitionId

But if that index already contains data (dicom Attributes indexed on dataset ids) I don't want to overwrite it, I want to merge it, adding more (datasetId => Attributes) to the given acquisitionId.